### PR TITLE
Heading tag (fixes #5)

### LINKED ISSUE #5: Implement {heading} tag
Implement the existing tag used in the CSFG. The function is found within [`markdownsection.py`](https://github.com/uccser/cs-field-guide/blob/develop/generator/markdownsection.py).

```
[heading]
    regex: (?P<heading_level>^#{2,6}|\A#) ?(?P<heading>[^\n#]+)
    function: create_heading
```

This function did create a tree of heading nodes, and this will still need to be stored in Kordac for the Django system.

### PR DESCRIPTION
**Pull request by ravenmaster001**
fixes #5 
This is about ~700 additions take your time.

Includes:
  - Heading Tag
  - Heading Tree
  - Usage Docs update
  - Unique Slugify
  - Minor fixes to tests

@JackMorganNZ Can you look at this and see if this is what you wanted.

### CODE REVIEW

File: kordac/Kordac.py [hunk 1]
```diff
@@ -60,9 +61,10 @@ def convert(self, text):
         self.kordac_extension.clear_saved_data()
         html_string = self.converter.convert(text)
         result = KordacResult(
-            html_string=html_string,
-            title=self.kordac_extension.title,
-            required_files=self.kordac_extension.required_files
+            html_string = html_string,
+            title = self.kordac_extension.title,
+            required_files = self.kordac_extension.required_files,
+            heading_tree = self.kordac_extension.get_heading_tree()
```
Comment (hayleyavw) [inline kordac/Kordac.py]: @JackMorganNZ tends to use the style without spaces...should probably keep things consistent...or change ALL THE THINGS

File: kordac/processors/HeadingBlockProcessor.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        level_trail = self.level_generator.next(level)
+
+        context = dict()
+        context['heading_level'] = level
+        context['heading_type'] = "h{0}".format(level)
+        context['title'] = heading
+        context['title_slug'] = heading_slug
+        context.update(
+            zip(("level_{0}".format(level) for level in range(1, self.max_levels + 1))
+                , level_trail))
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: comma on new line? this looks gross

File: kordac/processors/HeadingBlockProcessor.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            True if the block matches the pattern regex of a HeadingBlock.
+        '''
+        return self.pattern.search(block) is not None
+
+    def run(self, parent, blocks):
+        ''' Increments the level trail at the given level. This also
+        zeros any levels after the incremented level.
+
+        Args:
+            level_inc: the level to be incremented.
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: specifying arguments that aren't there?...

File: kordac/processors/HeadingBlockProcessor.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+        Returns:
+            A tuple of levels, where higher levels are first and the lowest
+            level is last.
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: umm...

File: kordac/processors/HeadingBlockProcessor.py (hunk 3 above)
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: :scream:

File: kordac/processors/HeadingBlockProcessor.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            kwargs: Arguments handed to the super class.
+        '''
+        super().__init__(*args, **kwargs)
+        self.processor = 'heading'
+        self.max_levels = 6
+        self.pattern = re.compile(ext.processor_info[self.processor]['pattern'])
+        self.template = ext.jinja_templates[self.processor]
+        self.custom_slugify = ext.custom_slugify
+        self.level_generator = LevelGenerator(self.max_levels)
+
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: spontaneous new line?

File: kordac/processors/HeadingBlockProcessor.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+        Args:
+            level_inc: the level to be incremented.
+
+        Returns:
+            A tuple of levels, where higher levels are first and the lowest
+            level is last.
+        '''
+        block = blocks.pop(0)
+        match = self.pattern.search(block)
+        assert match is not None # If this is true how did we test successfully
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: is this line necessary (referencing your own comment) and do we have this check in all the other processors?

File: kordac/processors/HeadingBlockProcessor.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        heading_slug = self.custom_slugify(heading)
+        level_trail = self.level_generator.next(level)
+
+        context = dict()
+        context['heading_level'] = level
+        context['heading_type'] = "h{0}".format(level)
+        context['title'] = heading
+        context['title_slug'] = heading_slug
+        context.update(
+            zip(("level_{0}".format(level) for level in range(1, self.max_levels + 1))
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: can you explain how this works for me?

File: kordac/tests/HeadingTest.py [hunk 8]
```diff
@@ -13,3 +15,159 @@ def __init__(self, *args, **kwargs):
...
+        blocks = self.to_blocks(test_string)
+
+        self.assertListEqual([True, True, True], [HeadingBlockProcessor(self.ext, self.md.parser).test(blocks, block) for block in blocks], msg='"{}"'.format(test_string))
+
+        converted_test_string = markdown.markdown(test_string, extensions=[self.kordac_extension])
+        expected_string = self.read_test_file(self.processor_name, 'doc_example_basic_usage_expected.html', strip=True)
+        self.assertEqual(expected_string, converted_test_string)
+
+        tree = self.kordac_extension.get_heading_tree()
+        expected_tree = (HeadingNode(title='This is an H1',
```
Comment (hayleyavw) [inline kordac/tests/HeadingTest.py]: not sure if it seems terrible or not to have the expected tree inside the test functions, it just seems like it make the test file really long...but then I don't know if it makes sense to move it to it's own file...so I guess it's ok....hmmmm :thinking:

File: kordac/utils/UniqueSlugify.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        self.max_length = int(max_length)
+        self.word_boundary = bool(word_boundary)
+        self.separator = str(separator)
+        self.save_order = bool(save_order)
+        self.stopwords = tuple(stopwords)
+
+    def __call__(self, text):
+        '''
+        Args:
+            text: A string to be turned into a slug.
```
Comment (hayleyavw) [inline kordac/utils/UniqueSlugify.py]: and returns?...

File: kordac/utils/UniqueSlugify.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        '''
+        slug = slugify(text = text,
+                        entities = self.entities,
+                        decimal = self.decimal,
+                        hexadecimal = self.hexadecimal,
+                        max_length = self.max_length,
+                        word_boundary = self.word_boundary,
+                        separator = self.separator,
+                        save_order = self.save_order,
+                        stopwords = self.stopwords)
```
Comment (hayleyavw) [inline kordac/utils/UniqueSlugify.py]: those spaces again...get rid of 'em pls

**Comment by hayleyavw** [comment]
Also I assume there are more tests to come?.... Seems light on testing at the moment.....

File: kordac/processors/HeadingBlockProcessor.py (hunk 7 above)
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: after further discussion, I would like this changed to just use and ordinary for loop. please.
@JackMorganNZ input welcome.

File: kordac/processors/HeadingBlockProcessor.py (hunk 6 above)
Comment (ravenmaster001) [inline kordac/processors/HeadingBlockProcessor.py]: asserts like this are only active when python is in debug mode. We really should be asserting inputs in so many places but especially like this for other programmers, but that also takes time.

File: kordac/tests/HeadingTest.py (hunk 8 above)
Comment (ravenmaster001) [inline kordac/tests/HeadingTest.py]: If it is in a file then we even have to write a loader for them too, since they are really python objects.

Comment (hayleyavw) [inline kordac/tests/HeadingTest.py]: thought that would be the answer, they can stay!

File: kordac/processors/HeadingBlockProcessor.py [hunk 11]
```diff
@@ -75,16 +74,14 @@ def run(self, parent, blocks):
         context['heading_type'] = "h{0}".format(level)
         context['title'] = heading
         context['title_slug'] = heading_slug
-        context.update(
-            zip(("level_{0}".format(level) for level in range(1, self.max_levels + 1))
-                , level_trail))
-
+        for i, level_val in enumerate(level_trail):
+            context["level_{0}".format(i + 1)] = level_val
```
Comment (hayleyavw) [inline kordac/processors/HeadingBlockProcessor.py]: ahhhh, much better :stuck_out_tongue:

**Pull request merged**